lib/puppet/util.rb

```diff
@@ -478,13 +478,15 @@ def safe_posix_fork(stdin = $stdin, stdout = $stdout, stderr = $stderr, &block)
       $stderr = STDERR
 
       begin
-        Dir.foreach('/proc/self/fd') do |f|
-          if %{^\d+$}.match?(f) && f.to_i >= 3
-            begin
-              IO.new(f.to_i).close
-            rescue
-              nil
-            end
+        d = Dir.new('/proc/self/fd')
+        ignore_fds = ['.', '..', d.fileno.to_s]
+        d.each_child do |f|
+          next if ignore_fds.include?(f) || f.to_i < 3
```

Comment:
Looking at #41 I wonder if we should explicitly exclude the FDs that we want to keep open:
```suggestion
        ignore_fds = ['.', '..', stdin.fileno.to_s, stdout.fileno.to_s, stderr.fileno.to_s, d.fileno.to_s]
        d.each_child do |f|
          next if ignore_fds.include?(f)
```

This version is probably safer. Not sure if it's valid to have no stdin, stdout or stderr.
```suggestion
        ignore_fds = ['.', '..'] + [stdin, stdout, stderr, d].filter_map { |f| f&.fileno&.to_s }
        d.each_child do |f|
          next if ignore_fds.include?(f)
```

What makes me hesitant is that I don't know

Reply:
I'm a little worried we're playing whack-a-mole with things we want to filter out and I'm not sure how to be sure we're being exhaustive about it.

Reply:
Agree it's better than `< 3`. Just wonder if we have any other gaps of things we should be ignoring.

Reply:
I don't think so.

Ideally we would use `open3` instead of `fork()`. That can change the user, cwd, set env vars etc but that's a larger refactor I don't want to take on right now.